cli: split spinner into "Connecting" and "Initializing cluster" state for init (and prevent double initialization cases when an error is mistakenly retried)

Nirusu opened the pull request:
### Proposed change(s)
- Add "Connecting" spinner state for "constellation init"
- Show "Initializing cluster" only when our gRPC connection was successful 

This is based on @malt3 gRPC debug logging functions.
I figured we can also just use them for the normal user output :) 
Not sure if this is the best place to put it there, though since I am not sure if the state can theoretically change multiple times to READY. 
Hope I can get some feedback on that? Not super familiar with gRPC connection states and how this works in our init.
Otherwise it might make sense to make this a channel and let the caller handle this.

Second commit also adds a note that it can take a few minutes.

<!-- (uncomment if applicable)
### Related issue
- link to the issue
-->

<!-- (uncomment if applicable)
### Additional info
- Any additional information or context
-->

### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x], or check after submitting. -->

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [ ] Add labels (e.g., for changelog category)
- [ ] Link to Milestone

malt3 commented:
>  I am not sure if the state can theoretically change multiple times to READY.

Currently it certainly can change to READY multiple times. This is something we do want to prevent in the future by disallowing further init retries once we see that the connection became ready once.

malt3 commented on cli/internal/cmd/init.go:
```diff
@@ -262,6 +265,8 @@ func (d *initDoer) logGRPCStateChanges(ctx context.Context, wg *sync.WaitGroup,
 		}
 		if state == connectivity.Ready {
 			d.log.Debugf("Connection ready")
+			d.spinner.Stop()
+			d.spinner.Start("Initializing cluster ", false)
```
I think we should either prevent this spinner change from happening more than once in this func or prevent multiple init calls all together (see my other comment).

@daniel-weisse @3u13r any thoughts on preventing multiple init calls after the atls conn was established at least once?

daniel-weisse replied:
Dropping from an established connection to connecting should result in a failure imo.
We need to refactor the whole init logic anyway. Currently, a failed attestation will result in a warning about this being a non recoverable error (which is not true).

Nirusu replied:
Should we make a double `connectivity.Ready` a failure for now and discuss a larger refactor to init in an upcoming dev sync?

Because I would also +1 for a refactor just because of #1403 for which proper fixes also might include a refactor to init (from the bootstrapper side) to remove or better mitigate the pesky timeout logic.

daniel-weisse replied:
```shell
2023-03-14T08:39:56+01:00       DEBUG   cmd/init.go:246 Created protoClient
2023-03-14T08:39:56+01:00       DEBUG   cmd/init.go:259 Connection state started as CONNECTING
2023-03-14T08:39:56+01:00       DEBUG   cloudcmd/validators.go:170      Validating attestation document
2023-03-14T08:39:57+01:00       DEBUG   cloudcmd/validators.go:170      Successfully validated attestation document
2023-03-14T08:39:57+01:00       DEBUG   cmd/init.go:261 Connection state changed to CONNECTING
2023-03-14T08:39:57+01:00       DEBUG   cmd/init.go:264 Connection ready
// Init actually starts
```
Connection changes to `connecting` twice, so we should definitely not cancel here. But aborting with unretriable error after "Connection read" seems ok to me.

Nirusu replied:
Alright, I gave it a try with channels in the new commit. Check it out,  this makes things definitely more complicated though 😄

daniel-weisse commented on cli/internal/cmd/init.go:
```diff
@@ -209,21 +243,25 @@ func (i *initCmd) initCall(ctx context.Context, dialer grpcDialer, ip string, re
+	err := retrier.Do(initCtx)
+	if err != nil {
+		if errors.Is(err, context.Canceled) && connectedTwice {
+			return nil, &nonRetriableError{fmt.Errorf("init call: temporarily lost gRPC connection but resumption is not supported")}
```
```suggestion
			return nil, &nonRetriableError{errors.New("init call: temporarily lost gRPC connection but resumption is not supported")}
```

daniel-weisse reviewed (commented):
Overall implementation idea looks good to me, though I would like this to be extensively tested before merging.
Is it possible to write a unit test for this?

Nirusu commented:
Okay, now the unit test actually complain about data races with the use of the logger in the goroutine. Will investigate.

malt3 commented on cli/internal/cmd/init.go:
```diff
@@ -195,12 +196,45 @@ func (i *initCmd) initialize(cmd *cobra.Command, newDialer func(validator *cloud
+	// If yes, it switches the spinner to "Initializing cluster".
+	// If it happens multiple times, we cancel the context and abort.
+	// This usually means we lost the connection and opened a new session (which will likely fail).
+	var connectedTwice bool
```
```suggestion
	var connectedTwice bool
```

I think what you should to do here is really just use a single variable (`connected`), to check if a connection was established at least once.
We don't even want to try the second time.
So `connected` starts with the value `false`, if we get a signal from `connReadyChan`, we set `connected` to true and in the error handling below we should check if `connected` is true + we received an error and then return an unretriable error.

Nirusu replied:
I mainly used the second variable to differentiate the context was canceled from some other circumstance (e.g. pressing CTRL+C) during a "normal" connection (since I believe it is inherited from Cobra).

Nirusu replied:
Has been superseded by just using `context.CancelWithCause`.

Nirusu commented:
Added a test and removed that ugly `conntectedTwice` variable by just using `context.CancelWithCause`.

Now there's still two issues:
1. The data race still exists with the logger (effectively only reachable in the test I guess though). I don't know how to fix this.
2. Channels are blocking by default so I guess in a very unfortunate event `logGRPCStateChanges` could try to send on a blocking channel if the events change rapidly. I don't think that's really likely to happen, though?

This cancelling on multiple events is way more painful than I thought it would be... Might actually be easier to move this down to `logGRPCStateChanges` rather than using channels.

Nirusu commented:
Bye channels and bye data races, this should be easier if everything is just in the same function 😅

daniel-weisse commented on cli/internal/cmd/init.go:
```diff
@@ -288,8 +260,9 @@ func (d *initDoer) Do(ctx context.Context) error {
 }
 
-func (d *initDoer) logGRPCStateChanges(ctx context.Context, wg *sync.WaitGroup, conn *grpc.ClientConn) {
+func (d *initDoer) handleGRPCStateChanges(ctx context.Context, wg *sync.WaitGroup, conn *grpc.ClientConn) {
 	go func() {
```
Why does this function start a new Go routine instead of letting the caller handle this? @malt3

malt3 replied:
This is almost in line with how the rest of our code handles go routines. CC @katexochen 

The general pattern we currently follow:

```go
func foo(ctx context.Context) {
  var wg sync.WaitGroup
  defer wg.Wait()
  
  usesAGoRoutine(ctx, wg)
  // do something ....
}

func usesAGoRoutine(ctx context.Context, wg sync.WaitGroup) {
  wg.Add(1)
  go func() {
    defer wg.Done()
    // do something in the background
  }
}
```

Nirusu commented:
Actually thanks to the linter warning I see that this doesn't work as aspected given that the goroutine is shutdown when the state changes to READY. Will need to fix this still.

malt3 commented on cli/internal/cmd/init.go:
```diff
@@ -240,7 +251,8 @@ func (d *initDoer) Do(ctx context.Context) error {
 	grpcStateLogCtx, grpcStateLogCancel := context.WithCancel(ctx)
 	defer grpcStateLogCancel()
 	wg.Add(1)
```
nit (that i produced): wg.Add should be in `handleGRPCStateChanges`

daniel-weisse commented on cli/internal/cmd/init.go:
```diff
@@ -252,7 +263,8 @@ func (d *initDoer) Do(ctx context.Context) error {
 	return nil
 }
 
-func (d *initDoer) logGRPCStateChanges(ctx context.Context, wg *sync.WaitGroup, conn *grpc.ClientConn) {
+func (d *initDoer) handleGRPCStateChanges(ctx context.Context, wg *sync.WaitGroup, conn *grpc.ClientConn) {
+	wg.Add(1)
 	go func() {
```
I think this function would be a lot more readable if you moved the Go routine from being called in this function to the caller.
I.e  instead of doing `wg.Add(1)` and `go func() {` in `handleGRPCStateChanges`, do so in the calling function.
This would also have the benefit of making things a bit more clear to what is happening in the calling function.

Nirusu replied:
I personally agree with you but the others don‘t, so I don’t know ¯\\_(ツ)_/¯

However if we have this pattern already somewhere else in the code I would just go for consistency.

daniel-weisse replied:
@katexochen could you explain why you want to keep calling the go routine in the function?

katexochen replied:
It bundles concurrency related code in a single place, which makes it easier to reason about. There is a single place where mistakes could be made. If the caller needs to handle it, you'd need to check every caller for correct use. Functions should always aim to hide complexity like this from the caller, same goes for synchronization primitives like mutex etc.
